### DISCUSSION TITLE
Function behind Button not getting executed

### DISCUSSION THREAD

**Discussion by SirZeeno** [AvaloniaUI/Avalonia#15618] [Q&A] [answered]
````
<Button Grid.Column="1" x:DataType="importWindow:ImportWindow" Command="{Binding AddPathToQueue}" CommandParameter="{Binding ElementName=PathTextBox}">
                <StackPanel Orientation="Horizontal">
                    <ui:SymbolIcon Symbol="Add" FontSize="20" Foreground="White" HorizontalAlignment="Center"/>
                    <TextBlock Text="Add Path"/>
                </StackPanel>
            </Button>
````

````
public void AddPathToQueue(TextBox textBox)
    {
        ConsoleExt.WriteLineWithPretext("Added Path", ConsoleExt.OutputType.Info);
    }
````

I am trying to execute the function with the press of a button, and this way of executing worked for me before, the only different thing is that I am setting the data type for the button itself since the rest of the axaml relies on a different datatype. What I am expecting is a console output of "Added Path" but nothing is happening when I press the button.

To clarify, all the references in terms of bindings and data type are current and link up in my IDE and don't produce any compile errors or any errors when the button is pressed and this all takes place in a window that gets created when a button is pressed. All it produces is absolutely nothing, so any help is greatly appreciated in this. :)

**Comment by timunie** [comment]
I highly suggest to follow the MVVM pattern and use Commands in the ViewModel. Typically you do not pass View-items to the Command, but the containing text. Sample: 
https://github.com/AvaloniaUI/Avalonia.Samples/tree/main/src/Avalonia.Samples/MVVM/CommandSample

**Reply by thevortexcloud** [reply]
More specifically: You can't bind directly to methods. You need to bind to a command which then calls your method.

**Reply by SirZeeno** [reply]
So, I tried the way that was described in the sample, yet the button stays greyed out, so I can't even test if the works now. I set the canExecute to 

````
this.WhenAnyValue(vm => vm.PathTextBox, (name) => !string.IsNullOrEmpty(name));
`````

where PathTextBox is the text within the textbox and where it should turn to true when the text of the textbox is not null. The text box gets updated with

````
public string? PathTextBox
    {
        get => _inputPath;
        set => this.RaiseAndSetIfChanged(ref _inputPath, value);
    }
````

and the text within the textbox get bound to 

````
Text="{Binding PathTextBox}"
````

and the button command point to and gets a parameter of 

````
Command="{Binding AddPathToQueueCommand}" CommandParameter="{Binding PathTextBox}"
````

I'm starting not to understand why certain issues are happening and what they get caused by, so thank you for any help that you could provide in this. :)

**Reply by thevortexcloud** [reply]
Can you show your full command declaration and instantiation? Have you checked if it works without the `CanExecute` argument?

The button will also be disabled if your command binding is not set to a valid value (meaning the `AddPathToQueueCommand` value is possibly null).

**Reply by timunie** [reply]
check your DataContext. Most likely the command is null. Use DevTools F12 to explore that property.

**Reply by SirZeeno** [reply]
The button is greyed out even with the CanExecute removed, but the command is declared as specified by the sample

````
AddPathToQueueCommand = ReactiveCommand.Create<string?>(path => AddPathToQueue(path));
````

even the IDE is telling me that the AddPathToQueueCommand is being used by the axaml binding and when I set it to the ReactiveCommand.Create function.

**Reply by thevortexcloud** [reply]
My next guess is you are binding to the wrong thing or your datacontext is null. Are you binding to a list of items by any chance? Have you checked if your view model constructor is actually being called?

 You should be able to check the data context using the dev tools.

**Reply by SirZeeno** [reply]
the command and parameter in the f12 are null even though I am declaring them in the "public MainViewModel()" function, which should execute right away when the class gets called, shouldn't it?

**Reply by thevortexcloud** [reply, marked as answer]
A constructor will be called whenever the class is instantiated. If it's not being instantiated it wont be called. You can prove that the constructor is being called by putting a break point inside it with your IDE and seeing if it gets hit.

Avalonia generally does not try to instantiate your view models for you. You need to that yourself somehow (and in some cases explicitly set the view model as the data context for your view). The main window's view model is typically set in the `App.axaml.cs` file. Like this:

```csharp
    public override void OnFrameworkInitializationCompleted() {
        if (ApplicationLifetime is IClassicDesktopStyleApplicationLifetime desktop) {
            desktop.MainWindow = new MainWindow {
                DataContext = new MainWindowViewModel(),
            };
        }

        base.OnFrameworkInitializationCompleted();
    }
```

Child/other views can be created by instantiating view models and using a view locator to work out what view to load. Or you can just manually set it all up in your view's code behind. There are _many_ ways to structure it.

Without seeing more of your code it's hard to say what is going on here.

**Reply by SirZeeno** [reply]
Thank you, i just needed to set the Datacontext inside the creation of the window, which I didn't know you had to do.

**Reply by thevortexcloud** [reply]
You are welcome. I suggest you look into how the view locator works, which will auto create your views and assign the datacontext for you.

https://docs.avaloniaui.net/docs/concepts/view-locator
